spec/justified.tla

```diff
@@ -0,0 +1,249 @@
+    \* another pseudo-recursive computation
+    /\ is_ancestor_descendant_relationship(
+        get_block_from_hash(checkpoint.block_hash, node_state),
+        get_block_from_hash(vote.message.ffg_target.block_hash, node_state),
+        node_state
+        )
+    /\ is_ancestor_descendant_relationship(
+        get_block_from_hash(vote.message.ffg_source.block_hash, node_state),
+        get_block_from_hash(checkpoint.block_hash, node_state),
+        node_state
+        )
+
+\* With the predicate, we can filter out the votes that are relevant in one step
+\* @type: ($checkpoint, $commonNodeState) => Set($signedVoteMessage);
+VotesInSupportAssumingJustifiedSource(checkpoint, node_state) == 
+    { vote \in node_state.view_votes: IsVoteInSupportAssumingJustifiedSource(vote, checkpoint, node_state)}
+
+\* In the recursive implementation, computing is_justified_checkpoint(C) for a given checkpoint C
+\* requires us to look at the set of all votes in VotesInSupportAssumingJustifiedSource(C, ...)
+\* which additionally have source checkpoints satisfying is_justified_checkpoint recursively.        
+\* This naturally defines, for our choice of C, a set of (other) checkpoints, 
+\* for which we need to be able to tell whether or not they are justified.
+\* For certain checkpoints, such as for example the genesis checkpoint C_G, as well as 
+\* any checkpoints which do not have any votes in support, this set is empty.
+
+\* Each such checkpoint c_i pending justification requires us to look at VotesInSupportAssumingJustifiedSource(c_i, ...),
+\* which in turn gives us another (possibly empty) set of (other) checkpoints that need to be evaluated.
+\* We will show below that this construction necessarily terminates, but for now we note that:
+\*   - in each step, we have a set of relevant checkpoints
+\*   - for each of these checkpoints, we need to keep track of a set of votes potentially justifying it, 
+\*     which will be used to define the checkpoints in the next step
+
+\* Due to the above, we observe that we can define a (finite) sequence of _maps_ s.t.:
+\*   - The domain of the i-th map is exactly the set of checkpoints pending justification in the i-th step
+\*   - Each checkpoint in this domain maps to the set of votes used to potentially justify it
+
+\* Let CheckpointsPendingJustification_i denote the set of checkpoints we need to be able to justify in the i-th step
+\* Initially, CheckpointsPendingJustification_1 = { C }, i.e. the original checkpoint C
+\* To formalize the construction described above, CheckpointsPendingJustification_{i+1} will contain the sources of 
+\* all checkpoints, used by votes in support of any of the checkpoints in CheckpointsPendingJustification_i:
+\* CheckpointsPendingJustification_{i+1} ==
+\*     UNION { Sources(VotesInSupportAssumingJustifiedSource(c_j, ...)) : c_j \in CheckpointsPendingJustification_i }
```

Comment:
Is the `_j` here for "justified"? It's not an index... maybe we just call it `c`?